src/dmd/root/rmem.d

```diff
@@ -410,3 +410,78 @@ pure nothrow unittest
     string sEmpty;
     assert(sEmpty.arraydup is null);
 }
+
+//debug = Pool;
+//debug = PoolSummary;
+
+/**
+Defines a pool for class objects. Objects can be fetched from the pool with make() and returned to the pool with
+dispose(). Using a reference that has been dispose()d has undefined behavior. make() may return memory that has been
+previously dispose()d.
+
+Internally the implementation uses a singly-linked freelist with a global root. The "next" pointer is stored in the
+first word of each disposed object.
+*/
+struct Pool(T)
+if (is(T == class))
+{
+    /// The freelist's root
+    private static T root;
+
+    private static void trace(string fun, string f, uint l)()
+    {
+        debug(Pool)
+        {
+            debug(PoolSummary)
+            {
+                static ulong calls;
+                if (calls == 0)
+                {
+                    // Plant summary printer
+                    static extern(C) void summarize()
+                    {
+                        fprintf(stderr, "%.*s(%u): bytes: %lu calls: %lu Pool!(%.*s)."~fun~"()\n",
+                            cast(int) f.length, f.ptr, l, ((T.classinfo.initializer.length + 15) & ~15) * calls,
+                            calls, cast(int) T.stringof.length, T.stringof.ptr);
+                    }
+                    atexit(&summarize);
+                }
+                ++calls;
+            }
+            else
+            {
+                fprintf(stderr, "%.*s(%u): bytes: %lu Pool!(%.*s)."~fun~"()\n",
+                    cast(int) f.length, f.ptr, l, T.classinfo.initializer.length,
+                    cast(int) T.stringof.length, T.stringof.ptr);
+            }
+        }
+    }
+
+    /**
+    Returns a reference to a new object in the same state as if created with new T(args).
+    */
+    static T make(string f = __FILE__, uint l = __LINE__, A...)(auto ref A args)
+    {
+        if (!root)
+        {
+            trace!("makeNew", f, l)();
+            return new T(args);
+        }
+        trace!("makeReuse", f, l)();
+        auto result = root;
+        root = *(cast(T*) root);
+        memcpy(cast(void*) result, T.classinfo.initializer.ptr, T.classinfo.initializer.length);
+        result.__ctor(args);
+        return result;
+    }
+
+    /**
+    Signals to the pool that this object is no longer used, so it can recycle its memory.
+    */
+    static void dispose(string f = __FILE__, uint l = __LINE__, A...)(T goner)
+    {
+        trace!("dispose", f, l)();
```

Comment:
If `debug` is on, this should stomp on the object with `0xDEADBEEF`.

Reply:
Perhaps it should anyway, as that will prevent false pointers from being recognized by the GC.

Reply:
I changed the logic so the freelist is used ONLY if the GC is not used. In debug mode the memory is now stomped with `0xdeadbeef`.